src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -761,7 +761,7 @@ private string GetDebuggerDisplay()
                         bool isVar;
                         bool isConst = false;
                         AliasSymbol alias;
-                        var declType = BindVariableTypeWithAnnotations(component.Designation, diagnostics, component.Type.SkipScoped(out _).SkipRef(out _), ref isConst, out isVar, out alias);
+                        var declType = BindVariableTypeWithAnnotations(component.Designation, diagnostics, component.Type.SkipScoped(out _).SkipRef(), ref isConst, out isVar, out alias);
```

Comment:
here's teh general intuition (which i commented in the PR).  If a feature is ignoring 'ref', then it's a place where 'ref' is not legal (otherwise they wouldn't be able to ignore it). And those places already have a syntactic check to report that.  i.e. we already hve code on line 841 in this file to error for that case.  As such, we don't need to report a diagnostic about needing a particular lang version to use 'ref'.

However, if hte feature does care about ref-type and actually uses that ref-kind, then it is only legal to do that in a version of c# that supports `ref`.  So if it uses the actual RefKind, it needs to pass in diagnostics to get the right version checks.

Reply:
It seemed like in some places, we are skipping ref, but some other check is deciding conditionally whether ref is OK in that position. For example, in `BindVariableTypeWithAnnotations`. It looks like some of the time you deliberately called the overload to pass `diagnostics: null`, like in SourceLocalSymbol.

It looks like the `SkipRef` overload that takes a diagnostic bag is specifically checking for the *ref locals and returns* feature. I am wondering if the name of this method should be changed, e.g. `SkipRef()` with no parameters sticks around, and `SkipRef(DiagnosticBag, out RefKind` gets renamed to something like `SkipRefInLocalOrReturn(DiagnosticBag, out RefKind)`. Then it might be slightly easier to tell at a glance why it's expected to pass a diagnostic bag or not.